net/src/ipv4/mod.rs

```diff
@@ -0,0 +1,456 @@
+        /// you would eventually reach the set of all [`Ipv4`]
+        /// (as should be true with any implementation of [`Arbitrary`]).
+        ///
+        /// Unfortunately, the current implementation does not cover [`Ipv4::options`].
+        fn arbitrary(u: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
+            let mut header = Ipv4 {
+                inner: Ipv4Header::default(),
+            };
+            header.set_source(u.arbitrary()?);
+            header.set_destination(u.arbitrary()?);
+
+            // safety:
+            // safe in-so-far as
+            // 1. the entire point of this code is to test the integrity of the rest of the system
+            //    by generating untrustworthy headers.
+            // 2. this code is not shipped in production builds in the first place.
+            #[allow(unsafe_code)]
+            unsafe {
+                header.set_next_header(NextHeader::arbitrary(u)?.into());
+            }
+            // DSCP is a 6-bit field, so ignore the top two bits.
+            let dscp = Ipv4Dscp::try_new(u.arbitrary::<u8>()? & 0b0011_1111)
+                .unwrap_or_else(|e| unreachable!("{e:?}"));
+            // ECN is a 2-bit field, so ignore the top 6 bits.
+            let ecn = Ipv4Ecn::try_new(u.arbitrary::<u8>()? & 0b0000_0011)
+                .unwrap_or_else(|e| unreachable!("{e:?}"));
+            header.set_ttl(u.arbitrary()?);
+            header.set_dscp(dscp).set_ecn(ecn);
+            header.set_dont_fragment(u.arbitrary()?);
+            header.set_more_fragments(u.arbitrary()?);
+            header.set_identification(u.arbitrary()?);
+            // IP fragment offset is a 13-bit field, so ignore the top 3 bits.
+            let ip_frag_offset =
+                IpFragOffset::try_new(u.arbitrary::<u16>()? & 0b0001_1111_1111_1111)
+                    .unwrap_or_else(|e| unreachable!("{e:?}"));
+            header.set_fragment_offset(ip_frag_offset);
+            Ok(header)
+        }
+
+        fn size_hint(_depth: usize) -> (usize, Option<usize>) {
+            // When we start supporting the generation of arbitrary [`Ipv4::options`], this next
+            // line needs to be changed to `(Ipv4::MIN_LEN, Some(Ipv4::MAX_LEN))`
+            (Ipv4::MIN_LEN.get(), Some(Ipv4::MIN_LEN.get()))
+        }
+    }
+}
+
+#[allow(clippy::unwrap_used, clippy::expect_used)] // valid in test code
+#[cfg(test)]
+mod test {
+    use crate::ipv4::Ipv4;
+    use crate::parse::{DeParse, Parse, ParseError};
+    use etherparse::err::ipv4::{HeaderError, HeaderSliceError};
+
+    #[test]
+    fn parse_back() {
+        bolero::check!().with_arbitrary().for_each(|header: &Ipv4| {
+            let mut buffer = [0u8; Ipv4::MIN_LEN.get()];
+            let bytes_written = header
+                .write(&mut buffer)
+                .unwrap_or_else(|e| unreachable!("{e:?}"));
+            assert!(bytes_written <= Ipv4::MIN_LEN);
```

Comment:
Shouldn't this assertion just be assert_eq?

Reply:
Yup.  I think it was `assert!(bytes_written <= Ipv4::MAX_LEN)` in a prior version of the test and then the test was refactored and this happened.  Will fix